lib/howl/command.moon

```diff
@@ -182,48 +83,147 @@ command_completer = ->
+      log.error 'No such command'
+      return
+
+    @command_line\clear!
+    @command_line.prompt = markup.howl "<prompt>:</><commandname>#{cmdname}</> "
+
+    local results
+
+    ok, error = pcall ->
+      results = table.pack howl.app.window.command_line\run cmd
+
+    if not ok
+      log.error error
+
+    self.finish results and unpack results
+
+  on_update: (text) =>
+    if text\find ' '
+      @run_command text
+
+  command_completion: =>
+    text = @command_line.text
+    @command_line\clear!
+    cmdname = interact.select_command
+      :text
+      title: 'Command'
+
+    if cmdname
+      @run_command cmdname
+
+  run_historical: =>
+    text = @command_line.text
+    @command_line\clear!
+    @command_line\write_spillover text
+    cmdname = interact.select_historical_command!
+    if cmdname
+      @run_command cmdname
+
+  keymap:
+    tab: => @command_completion!
+    up: => @run_historical!
+    ctrl_r: => @run_historical!
+    escape: => self.finish!
+    enter: =>
+      if @command_line.text
+        @run_command @command_line.text
       else
-        cmd_completer or= command_completer!
-        cmd_completer text
-
-  readline = howl.app.window.readline
-  text = cmd_string
-  text = "#{cmd_string} " if cmd_string and not cmd_string\contains ' '
-  value = readline\read ':', cmd_input, :text
-  if value
-    state\submit value
-
-return setmetatable { :register, :unregister, :alias, :run, :names, :get }, {
-  __index: (key) => commands[key] or accessible_names[key]
+        @command_completion!
+
+command_runner = {
+  name: 'command-runner'
+  factory: CommandRunner
 }
+
+run = (cmd_string = nil, opts={}) ->
+  if not howl.app.window
+    error "Cannot run command '#{cmd_string}', application not initialized. Try using the 'app-ready' signal.", 2
+
+  cmd_string = nil if cmd_string == 'run'
+  command_line = howl.app.window.command_line
+  if opts.submit
+    command_line\run_auto_submit command_runner, cmd_string: cmd_string, directory: opts.directory
+  else
+    command_line\run command_runner, cmd_string: cmd_string, directory: opts.directory
+
+howl.interact.register
+  name: 'select_command'
+  description: 'Selection list for all commands'
+  evade_history: true
+  handler: (opts={}) ->
+    with opts
+      .items = get_command_items!
+      .headers = { 'Command', 'Key binding', 'Description' }
+      .columns = {
+        { style: 'string' }
+        { style: 'keyword' }
+        { style: 'comment' }
+      }
+      .submit_on_space = true
+
+    result = interact.select opts
+
+    if result
+      return result.selection[1]
+
+get_command_history = ->
+  howl.app.window.command_line.command_history
+
+interact.register
+  name: 'select_historical_command'
+  description: 'Selection list for previously run commands'
+  evade_history: true
+  handler: ->
+    line_items = {}
+    for idx, item in ipairs get_command_history!
+      table.insert line_items, {idx, item, command: tostring(item)}
```

Comment:
`table.insert` is aliased as `append` in this file, so that could be used here as well. Also, some subjective style nit-picking; this function seems very compact to me. I personally would prefer empty lines before `result = [..]` and `if result`.